### ISSUE TITLE
`slice::element_offset` can produce false positives via current operational semantics

### ISSUE THREAD

**Issue by Plecra** [rust-lang/rust#161376]
As of https://github.com/rust-lang/rust/pull/126770

https://github.com/rust-lang/rust/blob/d6080a1e2f79a051eeba6ca543672a7ea3d3f8cf/library/core/src/slice/mod.rs#L4573-L4584

the element_offset method is implemented by comparing the addresses of two pointers. For elements within the same allocation, this has the correct behaviour, however address equality and ordering between independent allocations have no guarantees via LLVM and I expect the intended rust opsem. For example, we want to be able to alias immutable allocations with the same data, and we want to be able to elide mallocs that are detected to not escape a scope, and the compiler invents addresses for the objects in that malloc: This is capable of producing pointers to independent live allocations with the same address.

With pointer comparison defined to ignore provenance, it cannot observe this distinction so can give a false positive, identifying an arbitrary element in a slice as having the "same address" as a pointer in a different allocation.

An example of a false positive would be 

```rs
fn foo(x: &[u32]) -> &u32 {
    assert!(x.len() > 0);
    x.element_offset(&10).unwrap()
}
// it is legitimate to transform this into 
fn foo(x: &[u32]) -> &u32 {
    assert!(x.len() > 0);
    // decide that the address of the temporary is == x.as_ptr().addr(), and simplify
    &x[0]
}
```

(edit for clarification: I'm not aware of rust's semantics here, I just know that in the past llvm has had independent allocations be incomparable in this way, and while talking with nia we found a series of examples of llvm treating them as such)

**Comment by nia-e** [comment]
i think @rust-lang/opsem is the right cc here; i believe our memory model allows this transformation, in which case libs needs to adjust the docs to say element_offset is nondeterministic (instead of returning a None) if the reference is from a different allocation? I'd very much like to just say allocations must always be strictly disjoint, but :/

**Comment by steffahn** [comment]
To me, this possibility of eliding malloc despite the need to then invent addresses and pass them to code that can observes an overlap would seem like a broken "optimization" which shouldn't actually be allowed to happen like that. That being said, I'm not too knowledgable on the relevant language semantics here so..

.. still, if this truly *is* an issue: then what would stop e.g. equality comparison on `Rc` (which also [compares pointers](https://doc.rust-lang.org/1.97.0/src/alloc/rc.rs.html#2630-2632) as a fast pass) from exibiting false positives as well?

**Comment by theemathas** [comment]
My understanding of the opsem is: invented addresses must always be observed to be non-overlapping with any existing allocation. So, this should not be a problem, right?

**Comment by nia-e** [comment]
I'll also note that if I understood this correctly, LLVM *does* perform this optimization and it's also the root cause of #107975

**Comment by Plecra** [comment]
Did some more reading around: This has been discussed in opsem over in this issue:

https://github.com/rust-lang/unsafe-code-guidelines/issues/328#issuecomment-1104374319
https://github.com/rust-lang/unsafe-code-guidelines/issues/328#issuecomment-1109535440

They're talking about exactly the ptr_eq issue :)

that issue links https://github.com/rust-lang/miri/issues/3450#issuecomment-2038630757 where there was more discussion of this

ah and https://github.com/rust-lang/unsafe-code-guidelines/issues/328#issuecomment-2741657300 is a fantastic summary of the state of that conversation

**Comment by RalfJung** [comment]
This thread confuses a bunch of separate questions and discussions.

> For elements within the same allocation, this has the correct behaviour, however address equality and ordering between independent allocations have no guarantees via LLVM and I expect the intended rust opsem. For example, we want to be able to alias immutable allocations with the same data, and we want to be able to elide mallocs that are detected to not escape a scope, and the compiler invents addresses for the objects in that malloc: This is capable of producing pointers to independent live allocations with the same address.

Ordering of allocations being unspecified is entirely fine -- that's normal non-determinism, nothing interesting or problematic going on. The only potential problem here is if different allocations overlap. Outside of LLVM bugs, I am not aware of any case where LLVM actually makes that happen in an observable way. So please bring evidence for your claim that LLVM can produce "independent live allocations with the same address".

https://github.com/rust-lang/unsafe-code-guidelines/issues/328 is about a theoretical problem where we have no good justification for the optimization that removes unused allocations. We have a way of formally justifying it, but that approach involves saying that allocations can arbitrarily overlap and we use pointer provenance to disambiguate. That causes problems for comparison fast-paths such as the one in slice equality. This is considered unacceptable which is why we have not pushed for that approach as a way of solving the issue with these optimizations. Further down that issue we are discussing ways to adjust the semantics of `==` on pointers to make it still behave as expected. There might be a workable approach there, though it may also impact optimizations, and we don't know how big that impact would be.

> I'll also note that if I understood this correctly, LLVM does perform this optimization and it's also the root cause of https://github.com/rust-lang/rust/issues/107975

That's one of those LLVM bugs that I mentioned above.

**Comment by Plecra** [comment]
"address equality and ordering between independent allocations have no guarantees" is the same as saying that different allocations can overlap yeah, but I think I was wrong about LLVM. It does seem to actually require disjoint allocation addresses in the integral addressspace (Just from a brief skim of discussions in their issue tracker today. I couldnt find somewhere that the langref commits strongly to what addresses actually are)

Assuming that `addr` on pointers in live allocations in rust are definitely disjoint then yup this seems entirely reasonable~
(Learning that overlapping allocations are being ruled out like this does have me thinking it'd be nice to avoid somehow. It seems worth considering different intrinsics for cases like the slice equality fast path - `definitely_equal_reads(p1,p2)` can capture the necessary semantics for them without requiring stronger semantics for address values in general. in particular, this means that address escape doesnt have to be part of pointer escape analysis. Haven't read the threads in any depth though and this could easily already have been discussed)

**Comment by RalfJung** [comment]
> "address equality and ordering between independent allocations have no guarantees" means that different allocations can overlap

That is not usually what it means, no. The non-overlap guarantee is generally implied.

**Comment by RalfJung** [comment]
> Assuming that addr on pointers in live allocations in rust are definitely disjoint then yup this seems entirely reasonable~
(Learning that overlapping allocations are being ruled out like this does have me thinking it'd be nice to avoid somehow. It seems worth considering different intrinsics for cases like the slice equality fast path - definitely_equal_reads(p1,p2) can capture the necessary semantics for them without requiring stronger semantics for address values in general. in particular, this means that address escape doesnt have to be part of pointer escape analysis. Haven't read the threads in any depth though and this could easily already have been discussed)

If this was a realistic option, I'd love to pursue it. It would fix what is probably the largest remaining open problem with the Rust AM semantics. But I am not sure that we can get people used to writing low-level software to accept "two pointers may have equal addresses and both be valid but read different data".

OTOH we do have to get them to accept that two function pointers may be equal and yet call different functions... see https://github.com/rust-lang/rust/issues/160202. So maybe then doing the same for data isn't that far off?

**Comment by tmiasko** [comment]
Allocations backing live variables are distjoint. Nevertheless some kinds of allocations are allowed to overlap, for example promoted expressions:

```rust
fn main() {
    let a = &[10];
    println!("{:?}", a.element_offset(&10));
}
```

```console
$ rustc a.rs && ./a
Some(0)
```

**Comment by RalfJung** [comment]
Promoteds can be deduplicated, i.e. multiple promoteds can become a single allocation. That's not the same as two distinct allocations overlapping. At least that's how I thought of it. Maybe saying that these are two distinct read-only allocations with the same content at the same address is easier? Not sure.

**Comment by asquared31415** [comment]
This seems analogous to how there can be two distinct functions that have the same address.

**Comment by RalfJung** [comment]
It is similar, yes. I linked https://github.com/rust-lang/rust/issues/160202 above.

But for functions (and vtables) we do our best to tell people "comparing function pointers is weird". We would have to do this for *all* pointers if we went through with the idea that all allocations can overlap.
